[webkitpy] Make SimulatedDevice.is_usable check more accurate

Smackteo opened the pull request:
#### 38e1eb990499819b374de736f5ab89701f8ae1db
<pre>
[webkitpy] Make SimulatedDevice.is_usable check more accurate
<a href="https://bugs.webkit.org/show_bug.cgi?id=244108">https://bugs.webkit.org/show_bug.cgi?id=244108</a>

Reviewed by Ryan Haddad.

* Tools/Scripts/webkitpy/xcode/simulated_device.py:
(SimulatedDevice):
(SimulatedDevice.is_usable):

Canonical link: <a href="https://commits.webkit.org/253846@main">https://commits.webkit.org/253846@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc7d17d0b35f17b64306a4745a56667145dd5204

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87081 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17929 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96054 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149662 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91057 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29533 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25799 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79214 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91108 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92697 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23850 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73901 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23802 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/90689 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78834 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79102 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66813 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27261 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12940 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27205 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13954 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2707 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28889 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36804 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33231 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

ryanhaddad commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -595,9 +595,10 @@ def is_usable(self, force_update=False):
             return True
-
-        exit_code = self.executive.run_command([SimulatedDeviceManager.xcrun, 'simctl', 'spawn', self.udid, 'launchctl', 'list', service], return_exit_code=True)
+        for service in service:
```
Is there a reason you're iterating through multiple services here? We only have one defined per device type.

Smackteo replied:
I sent this to Jonathan - I could technically drop the for loop (Though leaving it allows for additional apps in the future and whatnot.)... So we could get rid of it; but leaving it allows for additional options if someone wants to modify the method in the future. Either is fine with me.

JonWBedard replied:
Seeing this in total, I think getting rid of the loop is probably best.

If you wanted to retain the loop, the terminate would need to be in the loop, then the early return after the loop. But it's even a bit weirder than that, because in the loop case, you all listed applications to start, so it's likely that you're going to end up starting some services multiple times. If you were doing the loop, you should also early return on error. ie, if the first process failed to start, then we shouldn't even try the second.

ryanhaddad commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -595,9 +595,10 @@ def is_usable(self, force_update=False):
+            exit_code = self.executive.run_command([SimulatedDeviceManager.xcrun, 'simctl', 'launch', self.udid, service], return_exit_code=True)
         if exit_code == 0:
+            self.executive.run_command([SimulatedDeviceManager.xcrun, 'simctl', 'terminate', self.udid, service], return_exit_code=True)
             return True
```
If you do have a reason for iterating through multiple services, it seems like this `if` condition needs to be inside the `for` block.

Smackteo replied:
Terminating the app that's on the screen will send us back to the Home Screen so the test runner can be installed. This terminates whatever app has been launched; if we wanted to terminate each service we could put it in the for loop; should work either way, but I think dropping the loop and leaving this here would be better.

Smackteo committed:
[webkitpy] Make SimulatedDevice.is_usable check more accurate
https://bugs.webkit.org/show_bug.cgi?id=244108

Reviewed by Ryan Haddad.

* Tools/Scripts/webkitpy/xcode/simulated_device.py:
(SimulatedDevice):
(SimulatedDevice.is_usable):

Canonical link: https://commits.webkit.org/253846@main
